File: ws/src/robot_firmware/robot_firmware/simple_serial_transmitter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import serial

class SimpleSerialTransmitter(Node):
    def __init__(self):
        super().__init__("simple_serial_transmitter")
        self.declare_parameter("port", "/dev/ttyACM0")
        self.declare_parameter("baud_rate", 115200)
        self.sub = self.create_subscription(String, 'serial_transmitter', self.msgCallback, 10)
        
        self.port = self.get_parameter("port").value
        self.baud_rate: int = self.get_parameter("baud_rate").value
        self.arduino_ = serial.Serial(port = self.port, baudrate=self.baud_rate, timeout=0.1)


    def msgCallback(self, msg: String):
        self.get_logger().info(f"New message received, publishing on serial {self.arduino_.name}")
        self.arduino_.write(msg.data.encode())

def main():
    rclpy.init()
    simple_serial_transmitter = SimpleSerialTransmitter()
    rclpy.spin(simple_serial_transmitter)
    simple_serial_transmitter.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ws/src/robot_firmware/src/simple_serial_transmitter.cpp
#include <functional>
#include <libserial/SerialPort.h>
#include <libserial/SerialPortConstants.h>
#include <memory>
#include <rclcpp/executors.hpp>
#include <rclcpp/logging.hpp>
#include <rclcpp/node.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/subscription.hpp>
#include <rclcpp/utilities.hpp>
#include <std_msgs/msg/string.hpp>
#include <string>

using std::placeholders::_1;

class SimpleSerialTransmitter : public rclcpp::Node {
public:
  SimpleSerialTransmitter() : Node("simple_serial_transmitter") {
    declare_parameter<std::string>("port", "/dev/ttyACM0");
    std::string port_ = get_parameter("port").as_string();
    sub_ = create_subscription<std_msgs::msg::String>(
        "serial_transmitter", 10,
        std::bind(&SimpleSerialTransmitter::msgCallback, this, _1));

    arduino_.Open(port_);
    arduino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);
  }

  ~SimpleSerialTransmitter() { arduino_.Close(); };

  void msgCallback(const std_msgs::msg::String &msg) {
    RCLCPP_INFO_STREAM(this->get_logger(),
                       "New message received, publishing on serial port"
                           << msg.data);
    arduino_.Write(msg.data);
  }

private:
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
  LibSerial::SerialPort arduino_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleSerialTransmitter>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}